### PR TITLE
Add private chatrooms

### PR DESCRIPTION
**Pull request by marksho**
Add private chatrooms with the ability to add people.

### CODE REVIEW

File: src/main/java/codeu/controller/ChatServlet.java [hunk 1]
```diff
@@ -95,12 +98,39 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
     List<Message> messages = messageStore.getMessagesInConversation(conversationId);
+    List<User> users = new ArrayList<>();
+    users.addAll(userStore.getUsers());
+    Set<UUID> existingUsersId = conversation.getUsers();
+    for (Iterator<User> iter = users.iterator(); iter.hasNext(); ) {
+      User user = iter.next();
+      if (existingUsersId.contains(user.getId())) {
+        iter.remove();
+      }
+    }
```
Comment (rjuang) [inline src/main/java/codeu/controller/ChatServlet.java]: If you use a Set, i.e.,  `Set<User> users = new HashSet<>()`
You could simplify the for loop as follows:
```
   Set<User> users = new HashSet<>();
   users.addAll(userStore.getUsers());
   Set<User> existingUsersId = conversation.getUsers();
   users.removeAll(existingUsersId);
```

File: src/main/java/codeu/model/data/Conversation.java [hunk 2]
```diff
@@ -35,11 +37,12 @@
    * @param title the title of this Conversation
    * @param creation the creation time of this Conversation
    */
-  public Conversation(UUID id, UUID owner, String title, Instant creation) {
+  public Conversation(UUID id, UUID owner, String title, Instant creation, Set<UUID> users) {
```
Comment (rjuang) [inline src/main/java/codeu/model/data/Conversation.java]: Be careful of passing in references to collections and simply referencing the collection. The reason is because you can get into unexpected states as follows:
```
Set<UUID> myUsers = ... ;
Conversation conversationA = new Conversation(idA, ownerA, titel, creation, myUsers);
Conversation conversationB = new Conversation(idA, ownerA, titel, creation, myUsers);

conversationB.addUserId(userC);
```
Both `conversationA` and `conversationB` are mutated to include `userC`

I recommend you create the set inside the class and add the elements to the provided set instead. This also makes performance more predictable. What happens if the provided Set was implemented as a LinkedList? Then the lookups into the set would no longer be O(1).

To be more general, you could also make the constructor accept a `Collection<User>`.

tl;dr. Something like this:
```
private final Set<UUID> users = new HashSet<>();

public Conversation(UUID id, UUID owner, String title, Instant creation, Collection<UUID> users) {
    ...
    this.users.addAll(users);
}
```

File: src/main/java/codeu/model/store/basic/ConversationStore.java [hunk 3]
```diff
@@ -88,6 +88,39 @@ public boolean loadTestData() {
     return conversations;
   }
 
+  /** Access the current set of public conversations known to the application. */
+  public List<Conversation> getAllPublicConversations() {
+    List<Conversation> publicConversations = new ArrayList<>();
```
Comment (rjuang) [inline src/main/java/codeu/model/store/basic/ConversationStore.java]: (Optional) If Java 8 is supported (which I think it is), you can take advantage of streams and make this more readable as follows:
```
return conversations.stream()
  .filter(Conversation::isPrivate)
  .collect(Collectors.toList());
```

Also, you can do the same with below loops.

File: src/main/java/codeu/model/store/basic/DefaultDataStore.java [hunk 4]
```diff
@@ -118,7 +118,7 @@ private void addRandomConversations() {
       User user = getRandomElement(users);
       String title = "Conversation_" + i;
       Conversation conversation =
-          new Conversation(UUID.randomUUID(), user.getId(), title, Instant.now());
+          new Conversation(UUID.randomUUID(), user.getId(), title, Instant.now(), null);
```
Comment (rjuang) [inline src/main/java/codeu/model/store/basic/DefaultDataStore.java]: maybe pass an empty collection instead of null ? i.e., use `ImmutableSet.of()`

File: src/main/java/codeu/controller/ChatServlet.java (hunk 1 above)
Comment (marksho) [reply src/main/java/codeu/controller/ChatServlet.java]: Users are many 'User's but existingUsersId are many 'UUID's, is there still a simpler way to write it?

Comment (rjuang) [reply src/main/java/codeu/controller/ChatServlet.java]: Oh i see. how about:
```
Set<UUID> existingUsersId = conversation.getUsers();
List<User> users = userStore.getUsers().stream()
   .filter(user -> !existingUsersId.contains(user.getId()))
   .collect(Collectors.toList());
```
